### PR TITLE
DA-641 mark obsolete organizations

### PR DESCRIPTION
**Pull request by Seeker1911**

### CODE REVIEW

File: rest-api/tools/import_organizations.py [hunk 1]
```diff
@@ -147,6 +147,19 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
 
       if hpo_id_list and not dry_run:
         self.delete_sql_statement(session, hpo_id_list)
+        self.hpo_dao._invalidate_cache()
+
+      for hpo in hpo_id_list:
+        # checking for entities again to know if we should mark them as obsolete
+        old_hpo = self.hpo_dao.get(hpo)
```
Comment (calbach) [inline rest-api/tools/import_organizations.py]: nit: would probably be better to fetch these all at once, then iterate.

File: rest-api/tools/import_organizations.py [hunk 2]
```diff
@@ -147,6 +147,19 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
+        if old_hpo:
+          log_prefix = '(dry run) ' if dry_run else ''
```
Comment (calbach) [inline rest-api/tools/import_organizations.py]: Pull out log_prefix above

File: rest-api/tools/import_organizations.py [hunk 3]
```diff
@@ -147,6 +147,19 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
+          logging.info(log_prefix + 'Marking old HPO as obsolete referenced in other table: %s',
+                       old_hpo.name)
+          sql = """ UPDATE HPO
```
Comment (calbach) [inline rest-api/tools/import_organizations.py]: Didn't you just delete these HPOs above?

File: rest-api/model/organization.py [hunk 4]
```diff
@@ -15,3 +15,4 @@ class Organization(Base):
   hpoId = Column('hpo_id', Integer, ForeignKey('hpo.hpo_id'), nullable=False)
   # Sites belonging to this organization.
   sites = relationship('Site', cascade='all, delete-orphan', order_by='Site.googleGroup')
+  isObsolete = Column('is_obsolete', SmallInteger, default=0)
```
Comment (danqrodney) [inline rest-api/model/organization.py]: You should be able to use Enum(ObsoleteStatus) here. See:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/model/participant.py#L43

You'll need to change script.py.mako or the generated migrations to import the enum class, e.g.

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/alembic/script.py.mako#L18

File: rest-api/tools/import_organizations.py [hunk 5]
```diff
@@ -140,15 +141,27 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
       for hpo in hpos_to_remove:
         old_hpo = self.hpo_dao.get_by_name(hpo)
         if old_hpo:
-          log_prefix = '(dry run) ' if dry_run else ''
-          logging.info(log_prefix + 'Deleting old HPO no longer in Google sheet: %s', old_hpo.name)
           hpo_id_list.append(old_hpo.hpoId)
           self.deletion_count += 1
 
       if hpo_id_list and not dry_run:
+        for hpo in hpo_id_list:
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: We should be able to do this with a single SQL statement with an in clause?

File: rest-api/tools/import_organizations.py [hunk 6]
```diff
@@ -140,15 +141,27 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
+          old_hpo = self.hpo_dao.get(hpo)
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: I don't think we need to do this lookup again; we just looked them up in existing_hpos?

We do probably need logic above that won't bother to try to delete hpos again if they are already marked obsolete... that will get rid of the log spam of saying "we're going to delete X" over and over.

File: rest-api/tools/import_organizations.py [hunk 7]
```diff
@@ -201,16 +215,30 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
       org_id_list = []
       for org in orgs_to_remove:
-        log_prefix = '(dry run) ' if dry_run else ''
         logging.info(log_prefix + 'Deleting old Organization no longer in Google sheet: %s', org)
         old_org = self.org_dao.get_by_external_id(org)
         org_id_list.append(old_org.organizationId)
         self.deletion_count += 1
 
       if org_id_list and not dry_run:
+        for org in org_id_list:
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: See above; same here.

File: rest-api/tools/import_organizations.py [hunk 8]
```diff
@@ -323,15 +352,28 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
       site_id_list = []
       for site in sites_to_remove:
-        log_prefix = '(dry run) ' if dry_run else ''
         logging.info(log_prefix + 'Deleting old Site no longer in Google sheet: %s', site)
         old_site = self.site_dao.get_by_google_group(site)
         site_id_list.append(old_site.siteId)
         self.deletion_count += 1
 
       if site_id_list and not dry_run:
+        for site in site_id_list:
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: See above; same here.

File: rest-api/test/unit_test/dao_test/site_dao_test.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def test_update(self):
     created_site = self.site_dao.insert(site)
     new_site = Site(siteId=created_site.siteId, siteName='site2',
                     googleGroup='site2@googlegroups.com',
-                    mayolinkClientNumber=123456, hpoId=UNSET_HPO_ID)
+                    mayolinkClientNumber=123456, hpoId=UNSET_HPO_ID, isObsolete=0)
```
Comment (danqrodney) [inline rest-api/test/unit_test/dao_test/site_dao_test.py]: Did we need to add this?

File: rest-api/tools/import_organizations.py [hunk 10]
```diff
@@ -139,14 +139,29 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
       hpo_id_list = []
       for hpo in hpos_to_remove:
         old_hpo = self.hpo_dao.get_by_name(hpo)
-        if old_hpo:
-          log_prefix = '(dry run) ' if dry_run else ''
-          logging.info(log_prefix + 'Deleting old HPO no longer in Google sheet: %s', old_hpo.name)
+        if old_hpo and old_hpo.isObsolete != ObsoleteStatus.OBSOLETE:
           hpo_id_list.append(old_hpo.hpoId)
           self.deletion_count += 1
+        elif old_hpo.isObsolete == ObsoleteStatus.OBSOLETE:
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: Should be if old_hpo and old_hpo.isObsolete... I think.

File: rest-api/tools/import_organizations.py [hunk 11]
```diff
@@ -139,14 +139,29 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
+                     old_hpo.name)
+        str_list = ','.join([str(i) for i in hpo_id_list])
+
+        sql = """ UPDATE HPO
+            SET is_obsolete = 1
+            WHERE hpo_id in ({params})""".format(params=str_list)
+
+        session.execute(sql)
+
+        self.hpo_dao._invalidate_cache()
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: Might want to do this after delete_sql_statement?

File: rest-api/tools/import_organizations.py [hunk 12]
```diff
@@ -201,14 +216,28 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
       for org in orgs_to_remove:
-        log_prefix = '(dry run) ' if dry_run else ''
-        logging.info(log_prefix + 'Deleting old Organization no longer in Google sheet: %s', org)
         old_org = self.org_dao.get_by_external_id(org)
-        org_id_list.append(old_org.organizationId)
-        self.deletion_count += 1
+        if old_org and old_org.isObsolete != ObsoleteStatus.OBSOLETE:
+          org_id_list.append(old_org.organizationId)
+          self.deletion_count += 1
+        elif old_org.isObsolete == ObsoleteStatus.OBSOLETE:
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: Same as above, check that old_org is not None first.

File: rest-api/tools/import_organizations.py [hunk 13]
```diff
@@ -201,14 +216,28 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
...
+        str_list = ','.join([str(i) for i in org_id_list])
+        sql = """ UPDATE organization
+            SET is_obsolete = 1
+            WHERE organization_id in ({org_id_list})""".format(org_id_list=str_list)
+        session.execute(sql)
+
+        self.org_dao._invalidate_cache()
+        # Try to delete old orgs.
+        logging.info(log_prefix + 'Deleting old Organization no longer in Google sheet: %s', org)
+        self.delete_sql_statement(session, str_list)
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: Same as above, might do this before invalidating cache.

File: rest-api/tools/import_organizations.py [hunk 14]
```diff
@@ -323,14 +352,29 @@ def _cleanup_old_entities(self, session, row_list, dry_run):
     if sites_to_remove:
       site_id_list = []
       for site in sites_to_remove:
-        log_prefix = '(dry run) ' if dry_run else ''
         logging.info(log_prefix + 'Deleting old Site no longer in Google sheet: %s', site)
         old_site = self.site_dao.get_by_google_group(site)
+      if old_site and old_site.isObsolete != ObsoleteStatus.OBSOLETE:
         site_id_list.append(old_site.siteId)
         self.deletion_count += 1
+      elif old_site.isObsolete == ObsoleteStatus.OBSOLETE:
```
Comment (danqrodney) [inline rest-api/tools/import_organizations.py]: Same as above, checked for old_site is not None.

File: rest-api/tools/import_organizations.py (hunk 10 above)
Comment (Seeker1911) [reply rest-api/tools/import_organizations.py]: done

File: rest-api/tools/import_organizations.py (hunk 11 above)
Comment (Seeker1911) [reply rest-api/tools/import_organizations.py]: done

File: rest-api/tools/import_organizations.py (hunk 12 above)
Comment (Seeker1911) [reply rest-api/tools/import_organizations.py]: done

File: rest-api/tools/import_organizations.py (hunk 13 above)
Comment (Seeker1911) [reply rest-api/tools/import_organizations.py]: done

File: rest-api/tools/import_organizations.py (hunk 14 above)
Comment (Seeker1911) [reply rest-api/tools/import_organizations.py]: done

**Comment by Seeker1911** [comment]
Tested by removing multiple sites,orgs,awardees from local copy.

**Pull request merged**